### PR TITLE
Add websocket protocol with http-proxy support

### PR DESCRIPTION
**Pull request by virusvfv**
Hi again )
According previous talks I re-coded websocket branch again with your last ligolo-ng commits.
This commit adds websocket with http-proxy support.
So U can easier view this PR.
Thanks and regads !!

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -91,11 +97,24 @@ func main() {
 	if *serverAddr == "" {
 		logrus.Fatal("please, specify the target host user -connect host:port")
 	}
-	host, _, err := net.SplitHostPort(*serverAddr)
-	if err != nil {
-		logrus.Fatal("invalid connect address, please use host:port")
+
```
Comment (nicocha30) [inline cmd/agent/main.go:100]: Should use "net/url" instead of manual parsing: https://gobyexample.com/url-parsing

File: cmd/agent/main.go [hunk 2]
```diff
@@ -105,33 +124,53 @@ func main() {
 
 	for {
 		var err error
-		if *socksProxy != "" {
-			if _, _, err := net.SplitHostPort(*socksProxy); err != nil {
-				logrus.Fatal("invalid socks5 address, please use host:port")
-			}
-			conn, err = sockDial(*serverAddr, *socksProxy, *socksUser, *socksPass)
+		if strings.Contains(*serverAddr, "https://") ||
```
Comment (nicocha30) [inline cmd/agent/main.go:127]: Same comment, don't use manual parsing

File: cmd/agent/main.go [hunk 3]
```diff
@@ -169,3 +208,57 @@ func connect(conn net.Conn) error {
...
+	httpTransport = &http.Transport{
+		MaxIdleConns:    http.DefaultMaxIdleConnsPerHost,
+		TLSClientConfig: config,
+		Proxy:           http.ProxyURL(proxyUrl),
+	}
+
+	httpClient := &http.Client{Transport: httpTransport}
+	httpheader := &http.Header{}
+	httpheader.Add("User-Agent", useragent)
+	//Add your additional headers here
```
Comment (nicocha30) [inline cmd/agent/main.go:238]: remove TODO

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -30,10 +34,29 @@ type ControllerConfig struct {
 	DomainWhitelist []string
 }
 
+var wsconn net.Conn
+
 func New(config ControllerConfig) Controller {
 	return Controller{Network: "tcp", Connection: make(chan net.Conn, 1024), ControllerConfig: config, startchan: make(chan error), SelfCertCache: "ligolo-selfcerts"}
 }
 
+type myHttpServer struct {
```
Comment (nicocha30) [inline pkg/controller/controller.go:43]: Rename struct to more compliant name, like ligoloHttpServer

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -91,20 +114,57 @@ func (c *Controller) ListenAndServe() {
...
+				err = s.Serve(listener)
+			}()
+			if err != nil {
+				logrus.Error(err)
+			}
+			//manual waiting until handler got connection and global variable wsconn is set by http handler
+			//this not so gracefully but effective ))
+			for {
+				if wsconn != nil {
+					logrus.Infof("Got websocket connection %s", wsconn.RemoteAddr())
```
Comment (nicocha30) [inline pkg/controller/controller.go:142]: Maybe a little verbose? Debugf instead of Infof?

File: cmd/agent/main.go (hunk 1 above)
Comment (virusvfv) [reply cmd/agent/main.go:100]: Done in last commit

File: cmd/agent/main.go (hunk 2 above)
Comment (virusvfv) [reply cmd/agent/main.go]: Done in last commit

File: cmd/agent/main.go (hunk 3 above)
Comment (virusvfv) [reply cmd/agent/main.go:231]: Done in last commit

File: pkg/controller/controller.go (hunk 4 above)
Comment (virusvfv) [reply pkg/controller/controller.go]: Done in last commit

File: pkg/controller/controller.go (hunk 5 above)
Comment (virusvfv) [reply pkg/controller/controller.go]: Done in last commit

File: cmd/agent/main.go (hunk 3 above)
Comment (nicocha30) [reply cmd/agent/main.go:231]: And "Blublublu" just under 😄

Comment (virusvfv) [reply cmd/agent/main.go]: Done

**Comment by nicocha30** [comment]
Hey @virusvfv 

I pushed some changes in your code. Please test before I can merge.

**Pull request merged**